---
title: "SynMut: Tools for designing Synonymous Mutants in Virus Research"
author: "Haogao Gu"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{SynMut: Designing Synonymous Mutants for Viral Research}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Introduction
Synonymous mutations refer to the mutations in DNA/RNA sequences that cause
__no__ modification in the translated amino acid sequence. Most of the
synonymous mutations are also silent mutations as they do not have observable
effects on the organism's phenotype. Designing mutant sequences with synonymous
mutations is often applied in many biological studies as a method to control
some unwanted changes in the amino acid sequences.

The codon usage bias and dinucleotide usage bias are two genomic signatures of
different sequences, even for synonymous sequences. Characterizing the functions
of synonymous sequences with different codon usage bias or dinucleotide usage
bias help to study their impacts on various biological functions. In fact, this 
method has been applied in many virology researches.

`SynMut` provides tools for generating multiple synonymous DNA sequences of
different features (in particular codon/dinucleotide usage pattern). User can
also specify mutable regions in the sequences (this is particular useful as
there are some conserved regions in the genome which we would not want to
modify). This tool was originally designed for generating recombinant virus
sequences in influenza A virus to study the effects of different dinucleotide
usage and codon usage, yet the functions provided in this package can be generic to
a variety of other biological researches.

# Getting started
## Installation
\#TODO
```{r}
library("SynMut")
```

## Input data
We use the example data in the package. The fasta file contains the segment 7
and segment 8 DNA sequences of Influenza A/Brisbane/59/2007 (H1N1) (BR59). The
region file in csv format reads as a `data.frame` specifying the mutable
positions (in amino acid sense) correspond to the DNA sequences.
```{r}
filepath.fasta <- system.file("extdata", "example.fasta", package = "SynMut")
filepath.csv <- system.file("extdata", "target_regions.csv", package = "SynMut")
region <- read.csv(filepath.csv)
head(region)
```
The `input_seq` function takes either a system file in fasta format or a DNAStringSet object as input,
to construct a `regioned_dna` object that is used in the `SynMut` package.
```{r}
rgd.seq <- input_seq(filepath.fasta, region)
rgd.seq
```

## Access data
Various `get_` functions were used to get the data:

* Access the DNA sequences. This will return a DNAStringSet object (from
`Biostrings` package).
```{r}
get_dna(rgd.seq)
```
* Access the specified mutable region.
```{r}
# get_region return a list
str(get_region(rgd.seq))
```
* Get the codon usage / dinucleotide usage / nucleotide usage
```{r}
get_cu(rgd.seq)
get_du(rgd.seq)
get_nu(rgd.seq)
```
* We also provide functions to: 
    - get the codon usage frequency of synonymous codons: `get_freq`
    - get Relative Synonymous Codon Usage (rscu) of synonymous codons: `get_rscu`

# Generating random synonymous sequences
Generating radom synonymous mutations (in specific region if provided in the
`input_seq`), with optionally keeping or not the original codon usage bias.
```{r}
get_cu(codon_random(rgd.seq)) - get_cu(rgd.seq)
get_cu(codon_random(rgd.seq, keep = TRUE)) - get_cu(rgd.seq)
```
We can also specify the `n` parameter to control the proportion of the codons to
be mutate.
```{r}
get_cu(codon_random(rgd.seq, n = 0.5)) - get_cu(rgd.seq)
```


# Generating synonymous sequences with maximum/minimum usage of specific codon

# Generating synonymous sequences with maximum/minimum usage of specific dinucleotide

# Generating synonymous sequences mimicing a specific codon usage pattern